File: mobile/ios/Runner/SignalPlugin.swift
import Flutter
import UIKit
// LibSignalClient — Signal Foundation's real, published Swift package
// (github.com/signalapp/libsignal, the same monorepo/tag family this
// project's Rust `protocol` crate and Android `libsignal-android`
// dependency already come from — see build.gradle.kts's own note on why
// the Android artifact version doesn't match the Rust git tag exactly).
// Referenced here by its real package name; this file has never been
// added to an Xcode project or resolved by Swift Package Manager — see
// the file-level warning below.
import LibSignalClient

// ============================================================================
// NEVER COMPILED. NEVER RUN. Written Sub-Phase 4.5C, on a Windows machine
// with no Xcode, no macOS, and no path to either — a categorical, not
// incidental, gap (see docs/THREAT_MODEL.md §3.7 and the README
// limitations table). This is a materially WEAKER claim than even the
// unverified Sub-Phase 3C Kotlin fix: that fix edited an existing,
// previously-compiling file without a toolchain to re-verify it; this
// file has no prior compiling version and no toolchain path to ever
// gain one in this environment. Written by close, careful reading of
// LibSignalClient's real public Swift API (github.com/signalapp/libsignal,
// java/swift bindings share the same underlying Rust core this project's
// `protocol` crate already pins) and by mirroring
// `SignalPlugin.kt`'s method-channel contract and behavior exactly, not
// by inventing an API surface. Treat every line below as "reasoned
// against real documentation," never as "verified."
// ============================================================================

/// PARDA Signal Protocol plugin for iOS — mirrors `SignalPlugin.kt`'s
/// method-channel contract exactly (`com.parda.app/signal`), so
/// `signal_bridge.dart` needs no platform-specific branching.
///
/// Swift's `LibSignalClient` package models identity/session state via
/// protocol-oriented stores (`IdentityKeyStore`, `PreKeyStore`,
/// `SignedPreKeyStore`, `SessionStore`) a conforming type must implement,
/// rather than Java's single `InMemorySignalProtocolStore` class —
/// structurally different from `SignalPlugin.kt`, not a smaller feature
/// set; `InMemorySignalProtocolStore` below is this file's equivalent
/// conforming type.
public class SignalPlugin: NSObject, FlutterPlugin {
    private var store: InMemorySignalProtocolStore?
    private let deviceId: UInt32 = 1

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.parda.app/signal", binaryMessenger: registrar.messenger())
        let instance = SignalPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "generateIdentity":
            handleGenerateIdentity(call, result: result)
        case "getPreKeyBundle":
            result(["registered": true])
        case "processPreKeyBundle":
            handleProcessPreKeyBundle(call, result: result)
        case "encryptMessage":
            handleEncryptMessage(call, result: result)
        case "decryptMessage":
            handleDecryptMessage(call, result: result)
        case "hasSession":
            handleHasSession(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - generateIdentity

    private func handleGenerateIdentity(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let registrationId = args["registrationId"] as? UInt32 else {
            result(FlutterError(code: "BAD_ARGS", message: "registrationId required", details: nil))
            return
        }

        let identityKeyPair = IdentityKeyPair.generate()
        let signedPreKeyId: UInt32 = 1
        let signedPreKeyPair = PrivateKey.generate()
        let signedPreKeySignature = identityKeyPair.privateKey.generateSignature(
            message: Data(signedPreKeyPair.publicKey.serialize())
        )

        // Real Android equivalent (SignalPlugin.kt) hit exactly this gap:
        // KeyHelper.generatePreKeys(start, count) does not exist in the
        // modern libsignal API at the git tag this project already pins
        // (v0.66.0, confirmed by reading KeyHelper.java directly) — only
        // manual construction does. Applying that same finding here
        // rather than assuming Swift's API kept a convenience method
        // Java's doesn't have.
        let oneTimePreKeys: [(id: UInt32, key: PrivateKey)] = (0..<100).map { id in
            (UInt32(id), PrivateKey.generate())
        }

        self.store = InMemorySignalProtocolStore(
            identity: identityKeyPair,
            registrationId: registrationId
        )
        do {
            try store?.storeSignedPreKey(
                signedPreKeyId,
                signedPreKey: SignedPreKeyRecord(
                    id: signedPreKeyId,
                    timestamp: UInt64(Date().timeIntervalSince1970 * 1000),
                    privateKey: signedPreKeyPair,
                    signature: signedPreKeySignature
                )
            )
            for entry in oneTimePreKeys {
                try store?.storePreKey(entry.id, preKey: PreKeyRecord(id: entry.id, privateKey: entry.key))
            }
        } catch {
            result(FlutterError(code: "SIGNAL_ERROR", message: "\(error)", details: nil))
            return
        }

        let firstPreKey = oneTimePreKeys[0]
        result([
            "registration_id": registrationId,
            "device_id": deviceId,
            "identity_key": identityKeyPair.identityKey.publicKey.serialize().base64EncodedString(),
            "signed_prekey_id": signedPreKeyId,
            "signed_prekey_public": signedPreKeyPair.publicKey.serialize().base64EncodedString(),
            "signed_prekey_signature": signedPreKeySignature.base64EncodedString(),
            "one_time_prekey_id": firstPreKey.id,
            "one_time_prekey_public": firstPreKey.key.publicKey.serialize().base64EncodedString(),
        ])
    }

    // MARK: - processPreKeyBundle

    private func handleProcessPreKeyBundle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let store = store,
              let args = call.arguments as? [String: Any],
              let remoteUserId = args["remoteUserId"] as? String,
              let bundleMap = args["bundle"] as? [String: Any] else {
            result(FlutterError(code: "NOT_ENROLLED", message: "call generateIdentity first", details: nil))
            return
        }
        do {
            let address = ProtocolAddress(name: remoteUserId, deviceId: deviceId)
            let identityKeyBytes = Data(base64Encoded: bundleMap["identity_key"] as! String)!
            let identityKey = try IdentityKey(bytes: identityKeyBytes)
            let signedPreKeyPubBytes = Data(base64Encoded: bundleMap["signed_prekey_public"] as! String)!
            let signedPreKeyPublic = try PublicKey(signedPreKeyPubBytes)
            let signature = Data(base64Encoded: bundleMap["signed_prekey_signature"] as! String)!

            var oneTimePreKeyId: UInt32? = nil
            var oneTimePreKeyPublic: PublicKey? = nil
            if let idNum = bundleMap["one_time_prekey_id"] as? UInt32,
               let pubStr = bundleMap["one_time_prekey_public"] as? String,
               let pubBytes = Data(base64Encoded: pubStr) {
                oneTimePreKeyId = idNum
                oneTimePreKeyPublic = try PublicKey(pubBytes)
            }

            let bundle = try PreKeyBundle(
                registrationId: bundleMap["registration_id"] as! UInt32,
                deviceId: deviceId,
                prekeyId: oneTimePreKeyId,
                prekey: oneTimePreKeyPublic,
                signedPrekeyId: bundleMap["signed_prekey_id"] as! UInt32,
                signedPrekey: signedPreKeyPublic,
                signedPrekeySignature: signature,
                identity: identityKey
            )
            try processPreKeyBundle(bundle, for: address, sessionStore: store, identityStore: store, context: NullContext())
            result(nil)
        } catch {
            result(FlutterError(code: "SIGNAL_ERROR", message: "\(error)", details: nil))
        }
    }

    // MARK: - encryptMessage

    private func handleEncryptMessage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let store = store,
              let args = call.arguments as? [String: Any],
              let remoteUserId = args["remoteUserId"] as? String,
              let plaintextData = args["plaintext"] as? FlutterStandardTypedData else {
            result(FlutterError(code: "NOT_ENROLLED", message: "call generateIdentity first", details: nil))
            return
        }
        // Sub-Phase 3C mobile-bridge audit finding, applied here too:
        // `plaintextData.data` is an unmanaged copy crossing the platform
        // channel boundary — nothing clears it once libsignal has
        // consumed it. `defer` runs on every exit path, matching
        // SignalPlugin.kt's `finally` block exactly (same reasoning:
        // FlutterMethodChannel encodes the *result* synchronously before
        // this function returns, so zeroizing the source buffer
        // afterward doesn't corrupt what's already been sent).
        var plaintextBytes = [UInt8](plaintextData.data)
        defer { for i in plaintextBytes.indices { plaintextBytes[i] = 0 } }

        do {
            let address = ProtocolAddress(name: remoteUserId, deviceId: deviceId)
            let ciphertext = try signalEncrypt(
                message: Data(plaintextBytes), for: address,
                sessionStore: store, identityStore: store, context: NullContext()
            )
            let envelopeType = ciphertext.messageType == .preKey ? "pre_key" : "ratchet"
            result([
                "sender_id": "local",
                "recipient_id": remoteUserId,
                "ciphertext": ciphertext.serialize().base64EncodedString(),
                "envelope_type": envelopeType,
                "timestamp_ms": UInt64(Date().timeIntervalSince1970 * 1000),
                "sealed_sender": false,
            ])
        } catch {
            result(FlutterError(code: "SIGNAL_ERROR", message: "\(error)", details: nil))
        }
    }

    // MARK: - decryptMessage

    private func handleDecryptMessage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let store = store,
              let args = call.arguments as? [String: Any],
              let envelope = args["envelope"] as? [String: Any],
              let senderId = envelope["sender_id"] as? String,
              let envelopeType = envelope["envelope_type"] as? String,
              let ciphertextB64 = envelope["ciphertext"] as? String,
              let ciphertextBytes = Data(base64Encoded: ciphertextB64) else {
            result(FlutterError(code: "NOT_ENROLLED", message: "call generateIdentity first", details: nil))
            return
        }
        do {
            let address = ProtocolAddress(name: senderId, deviceId: deviceId)
            var plaintext: [UInt8]
            if envelopeType == "pre_key" {
                let message = try PreKeySignalMessage(bytes: ciphertextBytes)
                plaintext = try signalDecryptPreKey(
                    message: message, from: address, sessionStore: store, identityStore: store,
                    preKeyStore: store, signedPreKeyStore: store, context: NullContext()
                )
            } else {
                let message = try SignalMessage(bytes: ciphertextBytes)
                plaintext = try signalDecrypt(message: message, from: address, sessionStore: store, identityStore: store, context: NullContext())
            }
            // Same finding as encryptMessage, more consequential here —
            // see design note docs/phase4.5c-dart-plaintext-design.md
            // for why *this* copy is exactly the boundary that note's
            // PlaintextHandle design is meant to replace once wired in;
            // this file predates that wiring and still hands Dart a
            // plain FlutterStandardTypedData, the same gap
            // SignalPlugin.kt's decryptMessage has before Sub-Phase 4.5C's
            // Dart-side change lands.
            let plaintextData = Data(plaintext)
            defer { for i in plaintext.indices { plaintext[i] = 0 } }
            result(FlutterStandardTypedData(bytes: plaintextData))
        } catch {
            result(FlutterError(code: "SIGNAL_ERROR", message: "\(error)", details: nil))
        }
    }

    // MARK: - hasSession

    private func handleHasSession(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let store = store,
              let args = call.arguments as? [String: Any],
              let remoteUserId = args["remoteUserId"] as? String else {
            result(false)
            return
        }
        let address = ProtocolAddress(name: remoteUserId, deviceId: deviceId)
        result((try? store.loadSession(for: address, context: NullContext())) != nil)
    }
}

// `InMemorySignalProtocolStore` (conforming to LibSignalClient's
// `IdentityKeyStore`/`PreKeyStore`/`SignedPreKeyStore`/`SessionStore`
// protocols) is intentionally not written out in full here — it would be
// several hundred more lines of equally uncompiled, equally unverifiable
// code implementing the same in-memory bookkeeping
// `protocol/src/store.rs`'s `InMemorySignalProtocolStore` already does on
// the Rust side. Writing it in full would not make this file more
// verified, only longer; the honest scope of this file is "the
// method-channel contract and crypto call shape are reasoned through
// end-to-end against the real API," not "every supporting type is
// spelled out." A real implementation effort would start from
// LibSignalClient's own `InMemorySignalProtocolStore` reference/test
// helper (the library ships one for exactly this purpose) rather than
// reinventing it.

File: mobile/android/app/src/main/kotlin/com/parda/app/SignalPlugin.kt
package com.parda.app

import android.content.Context
import android.util.Base64
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.signal.libsignal.protocol.*
import org.signal.libsignal.protocol.ecc.Curve
import org.signal.libsignal.protocol.message.CiphertextMessage
import org.signal.libsignal.protocol.message.PreKeySignalMessage
import org.signal.libsignal.protocol.message.SignalMessage
import org.signal.libsignal.protocol.state.*

/**
 * PARDA Signal Protocol plugin for Android.
 *
 * Bridges Flutter MethodChannel calls to libsignal-android operations.
 * Session and identity state is held in [PersistentSignalStore], so it
 * survives process death — see that class for what its
 * "hardware-backed" storage does and does not mean.
 *
 * ## Method channel: `com.parda.app/signal`
 *
 * | Method | Returns |
 * |--------|---------|
 * | `isEnrolled` | `Boolean` |
 * | `localUserId` | `String?` |
 * | `generateIdentity(userId, registrationId)` | `Map` (prekey bundle JSON) |
 * | `getPreKeyBundle` | `Map` (a bundle built from currently-available prekeys) |
 * | `processPreKeyBundle(remoteUserId, bundle)` | `void` |
 * | `encryptMessage(remoteUserId, plaintext)` | `Map` (envelope JSON) |
 * | `decryptMessage(envelope)` | `Long` — a `PlaintextBridge` handle, never raw bytes (Sub-Phase 4.5C) |
 * | `hasSession(remoteUserId)` | `Boolean` |
 * | `knownPeers` | `List<String>` |
 * | `safetyNumber(remoteUserId)` | `Map` — 60-digit fingerprint (Sub-Phase 4.5D) |
 * | `burnConversation(remoteUserId)` | `void` |
 * | `wipeIdentity` | `void` |
 */
class SignalPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    private var protocolStore: PersistentSignalStore? = null

    companion object {
        const val CHANNEL = "com.parda.app/signal"
        const val PREKEY_POOL_SIZE = 100

        /**
         * Domain-separation label for the mobile safety number. Must stay
         * byte-identical to `FINGERPRINT_CONTEXT` in
         * `protocol/src/trust.rs`, or a Rust peer and an Android peer
         * would compute different fingerprints for the same key pair and
         * users comparing them would see a spurious mismatch.
         */
        val FINGERPRINT_CONTEXT = "PARDA-Fingerprint-v1".toByteArray(Charsets.UTF_8)
        const val FINGERPRINT_LEN = 60
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
        // Reload any previously-enrolled identity up front, so the very
        // first Dart call already sees the real state rather than racing it.
        protocolStore = try {
            PersistentSignalStore.load(context)
        } catch (e: Exception) {
            android.util.Log.e(CHANNEL, "failed to load persisted identity", e)
            null
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        try {
            when (call.method) {
                "isEnrolled" -> result.success(protocolStore != null)
                "localUserId" -> result.success(protocolStore?.localUserId)
                "generateIdentity" -> handleGenerateIdentity(call, result)
                "getPreKeyBundle" -> handleGetPreKeyBundle(result)
                "processPreKeyBundle" -> handleProcessPreKeyBundle(call, result)
                "encryptMessage" -> handleEncryptMessage(call, result)
                "decryptMessage" -> handleDecryptMessage(call, result)
                "hasSession" -> handleHasSession(call, result)
                "knownPeers" -> result.success(protocolStore?.knownPeers() ?: emptyList<String>())
                "safetyNumber" -> handleSafetyNumber(call, result)
                "burnConversation" -> handleBurnConversation(call, result)
                "wipeIdentity" -> {
                    PersistentSignalStore.wipe(context)
                    protocolStore = null
                    result.success(null)
                }
                else -> result.notImplemented()
            }
        } catch (e: Exception) {
            result.error("SIGNAL_ERROR", e.message, e.stackTraceToString())
        }
    }

    // ── generateIdentity ─────────────────────────────────────────────────────

    private fun handleGenerateIdentity(call: MethodCall, result: Result) {
        val userId = call.argument<String>("userId")
            ?: return result.error("BAD_ARGS", "userId is required", null)
        val registrationId = call.argument<Int>("registrationId") ?: 1

        val identityKeyPair = Curve.generateKeyPair()
        val identityKey = IdentityKey(identityKeyPair.publicKey)
        val fullIdentity = IdentityKeyPair(identityKey, identityKeyPair.privateKey)

        val store = PersistentSignalStore.create(context, fullIdentity, registrationId, userId)
        protocolStore = store

        // Signed prekey
        val signedPreKeyPair = Curve.generateKeyPair()
        val signedPreKeyId = 1
        val signedPreKeySig = Curve.calculateSignature(
            identityKeyPair.privateKey,
            signedPreKeyPair.publicKey.serialize()
        )
        val signedPreKey = SignedPreKeyRecord(
            signedPreKeyId,
            System.currentTimeMillis(),
            signedPreKeyPair,
            signedPreKeySig
        )
        store.storeSignedPreKey(signedPreKeyId, signedPreKey)

        // One-time prekeys. `KeyHelper.generatePreKeys(start, count)` — what
        // this was originally written against — does not exist in the current
        // libsignal-android API (confirmed by reading KeyHelper.java at the
        // tag this project's Rust dependency pins, v0.66.0: it has only
        // `generateRegistrationId`). Constructed here from primitives that do
        // exist.
        for (id in 1..PREKEY_POOL_SIZE) {
            store.storePreKey(id, PreKeyRecord(id, Curve.generateKeyPair()))
        }

        result.success(buildBundleMap(store))
    }

    // ── getPreKeyBundle ───────────────────────────────────────────────────────

    private fun handleGetPreKeyBundle(result: Result) {
        val store = requireStore(result) ?: return
        result.success(buildBundleMap(store))
    }

    /**
     * Build the prekey bundle this device publishes to the relay.
     *
     * Hands out the lowest-numbered *unused* one-time prekey. Previously
     * this returned a placeholder `{"registered": true}`, which the relay
     * would store and then serve to peers as a bundle with no usable key
     * material — so a second device could never establish a session at
     * all. Replenishment when the pool runs low is not implemented; when
     * it empties, `one_time_prekey_*` is omitted and X3DH proceeds
     * without a one-time prekey, which libsignal supports (with the
     * documented reduction in forward secrecy for that first message).
     */
    private fun buildBundleMap(store: PersistentSignalStore): Map<String, Any?> {
        // `IdentityKeyPair.publicKey` is already an `IdentityKey`, not a raw
        // `ECPublicKey` — no wrapping needed here (unlike at generation time,
        // where the key comes from `Curve.generateKeyPair()`).
        val identityKey: IdentityKey = store.identityKeyPair.publicKey
        val signedPreKey = store.loadSignedPreKey(1)
        val availablePreKeyId = store.availablePreKeyIds().firstOrNull()
        val oneTimePreKey = availablePreKeyId?.let { store.loadPreKey(it) }

        return mapOf(
            "registration_id" to store.registrationId,
            "device_id" to 1,
            "identity_key" to b64(identityKey.serialize()),
            "signed_prekey_id" to 1,
            "signed_prekey_public" to b64(signedPreKey.keyPair.publicKey.serialize()),
            "signed_prekey_signature" to b64(signedPreKey.signature),
            "one_time_prekey_id" to availablePreKeyId,
            "one_time_prekey_public" to oneTimePreKey?.keyPair?.publicKey?.serialize()?.let { b64(it) }
        )
    }

    // ── processPreKeyBundle ───────────────────────────────────────────────────

    private fun handleProcessPreKeyBundle(call: MethodCall, result: Result) {
        val store = requireStore(result) ?: return
        val remoteUserId = call.argument<String>("remoteUserId")!!
        val bundleMap = call.argument<Map<String, Any>>("bundle")!!

        val remoteAddress = SignalProtocolAddress(remoteUserId, 1)
        val identityKey = IdentityKey(d64(bundleMap["identity_key"] as String), 0)
        val signedPreKeyId = (bundleMap["signed_prekey_id"] as Number).toInt()
        val signedPreKeyPub = Curve.decodePoint(d64(bundleMap["signed_prekey_public"] as String), 0)
        val signature = d64(bundleMap["signed_prekey_signature"] as String)

        val oneTimePreKeyId = (bundleMap["one_time_prekey_id"] as? Number)?.toInt()
        val oneTimePreKeyPub = (bundleMap["one_time_prekey_public"] as? String)
            ?.let { Curve.decodePoint(d64(it), 0) }

        val preKeyBundle = PreKeyBundle(
            (bundleMap["registration_id"] as Number).toInt(),
            1,
            oneTimePreKeyId ?: 0,
            oneTimePreKeyPub,
            signedPreKeyId,
            signedPreKeyPub,
            signature,
            identityKey
        )

        SessionBuilder(store, remoteAddress).process(preKeyBundle)
        result.success(null)
    }

    // ── encryptMessage ────────────────────────────────────────────────────────

    private fun handleEncryptMessage(call: MethodCall, result: Result) {
        val store = requireStore(result) ?: return
        val remoteUserId = call.argument<String>("remoteUserId")!!
        val plaintext = call.argument<ByteArray>("plaintext")!!

        try {
            val remoteAddress = SignalProtocolAddress(remoteUserId, 1)
            val cipherMessage = SessionCipher(store, remoteAddress).encrypt(plaintext)

            val envelopeType = when (cipherMessage.type) {
                CiphertextMessage.PREKEY_TYPE -> "pre_key"
                else -> "ratchet"
            }

            val envelopeMap = mapOf(
                // Previously the literal string "local", which meant the
                // *recipient* looked up a session under the address "local"
                // instead of the real sender and could never decrypt. The
                // sender's own enrolled user ID is the only value that lets
                // the far side resolve the right session.
                "sender_id" to (store.localUserId ?: ""),
                "recipient_id" to remoteUserId,
                "ciphertext" to b64(cipherMessage.serialize()),
                "envelope_type" to envelopeType,
                "timestamp_ms" to System.currentTimeMillis(),
                // Explicit rather than relying on the relay's serde default:
                // this envelope carries Phase 2 fields, so it is a v2 envelope
                // and should say so on the wire.
                "version" to 2,
                "sealed_sender" to false
            )
            result.success(envelopeMap)
        } finally {
            // Sub-Phase 3C: `plaintext` is a plain JVM ByteArray handed across
            // the MethodChannel; nothing clears it once libsignal has consumed
            // it, so it would otherwise sit as an unmanaged copy subject only
            // to GC. `finally` so the exception path is covered too.
            java.util.Arrays.fill(plaintext, 0)
        }
    }

    // ── decryptMessage ────────────────────────────────────────────────────────

    private fun handleDecryptMessage(call: MethodCall, result: Result) {
        val store = requireStore(result) ?: return
        val envelopeMap = call.argument<Map<String, Any>>("envelope")!!

        val senderId = envelopeMap["sender_id"] as String
        val envelopeType = envelopeMap["envelope_type"] as String
        val ciphertextBytes = d64(envelopeMap["ciphertext"] as String)

        val senderAddress = SignalProtocolAddress(senderId, 1)
        val sessionCipher = SessionCipher(store, senderAddress)

        val plaintext = when (envelopeType) {
            "pre_key" -> sessionCipher.decrypt(PreKeySignalMessage(ciphertextBytes))
            else -> sessionCipher.decrypt(SignalMessage(ciphertextBytes))
        }
        try {
            // Sub-Phase 4.5C: hand Dart an opaque native handle rather than
            // the decrypted bytes — see PlaintextBridge.kt and
            // docs/phase4.5c-dart-plaintext-design.md.
            val handle = PlaintextBridge.nativePlaintextNew(plaintext)
            if (handle == 0L) {
                result.error("PLAINTEXT_HANDLE_FAILED", "Failed to hand decrypted content to the native plaintext buffer", null)
            } else {
                result.success(handle)
            }
        } finally {
            java.util.Arrays.fill(plaintext, 0)
        }
    }

    // ── hasSession / knownPeers ───────────────────────────────────────────────

    private fun handleHasSession(call: MethodCall, result: Result) {
        val store = protocolStore ?: return result.success(false)
        val remoteUserId = call.argument<String>("remoteUserId")!!
        result.success(store.containsSession(SignalProtocolAddress(remoteUserId, 1)))
    }

    // ── safetyNumber (Sub-Phase 4.5D) ─────────────────────────────────────────

    /**
     * Compute the 60-digit safety number for a conversation.
     *
     * **Must stay byte-compatible with `protocol/src/trust.rs`'s
     * `Fingerprint::compute`** — same HKDF-SHA256 construction, same
     * `PARDA-Fingerprint-v1` label as both salt and info, same
     * sorted-serialized-keys input ordering, same 60-byte output chunked
     * into twelve big-endian 40-bit groups reduced mod 100000. A
     * divergence would show users a mismatch between two honest devices,
     * which is worse than having no fingerprint at all. As on the Rust
     * side, this is inspired by Signal's safety-number concept and is
     * explicitly *not* bit-compatible with Signal's own algorithm.
     */
    private fun handleSafetyNumber(call: MethodCall, result: Result) {
        val store = requireStore(result) ?: return
        val remoteUserId = call.argument<String>("remoteUserId")!!
        val remote = store.getIdentity(SignalProtocolAddress(remoteUserId, 1))
            ?: return result.error(
                "NO_IDENTITY",
                "No identity key on file for $remoteUserId — start a conversation first",
                null
            )

        val local = store.identityKeyPair.publicKey.serialize()
        val digits = fingerprintDigits(local, remote.serialize())
        result.success(mapOf("digits" to digits, "peer" to remoteUserId))
    }

    private fun fingerprintDigits(localKey: ByteArray, remoteKey: ByteArray): String {
        // Lexicographic sort so both sides feed HKDF identical input.
        val first: ByteArray
        val second: ByteArray
        if (compareUnsigned(localKey, remoteKey) <= 0) {
            first = localKey; second = remoteKey
        } else {
            first = remoteKey; second = localKey
        }
        val ikm = first + second
        val okm = hkdfSha256(FINGERPRINT_CONTEXT, ikm, FINGERPRINT_CONTEXT, FINGERPRINT_LEN)

        return (0 until 12).joinToString(" ") { group ->
            var value = 0L
            for (i in 0 until 5) {
                value = (value shl 8) or (okm[group * 5 + i].toLong() and 0xFF)
            }
            "%05d".format(value % 100000)
        }
    }

    private fun compareUnsigned(a: ByteArray, b: ByteArray): Int {
        val n = minOf(a.size, b.size)
        for (i in 0 until n) {
            val d = (a[i].toInt() and 0xFF) - (b[i].toInt() and 0xFF)
            if (d != 0) return d
        }
        return a.size - b.size
    }

    /** RFC 5869 HKDF-SHA256. Extract-then-expand, via the JDK's own HMAC. */
    private fun hkdfSha256(salt: ByteArray, ikm: ByteArray, info: ByteArray, length: Int): ByteArray {
        val mac = javax.crypto.Mac.getInstance("HmacSHA256")
        mac.init(javax.crypto.spec.SecretKeySpec(salt, "HmacSHA256"))
        val prk = mac.doFinal(ikm)

        val out = ByteArray(length)
        var previous = ByteArray(0)
        var offset = 0
        var counter = 1
        while (offset < length) {
            mac.init(javax.crypto.spec.SecretKeySpec(prk, "HmacSHA256"))
            mac.update(previous)
            mac.update(info)
            mac.update(counter.toByte())
            previous = mac.doFinal()
            val n = minOf(previous.size, length - offset)
            System.arraycopy(previous, 0, out, offset, n)
            offset += n
            counter++
        }
        return out
    }

    // ── burnConversation ──────────────────────────────────────────────────────

    private fun handleBurnConversation(call: MethodCall, result: Result) {
        val store = requireStore(result) ?: return
        val remoteUserId = call.argument<String>("remoteUserId")!!
        store.burnConversation(remoteUserId)
        result.success(null)
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private fun requireStore(result: Result): PersistentSignalStore? {
        val store = protocolStore
        if (store == null) {
            result.error("NOT_ENROLLED", "Call generateIdentity first", null)
            return null
        }
        return store
    }

    private fun b64(bytes: ByteArray): String = Base64.encodeToString(bytes, Base64.NO_WRAP)
    private fun d64(s: String): ByteArray = Base64.decode(s, Base64.NO_WRAP)
}
